src/main/webapp/app/exam/participate/exam-navigation-bar/exam-navigation-bar.component.ts

```diff
@@ -143,13 +143,11 @@ export class ExamNavigationBarComponent implements OnInit {
         if (submission && this.exercises[this.exerciseIndex].type !== ExerciseType.PROGRAMMING) {
             submission.submitted = true;
         }
-        if (changeExercise) {
-            if (newIndex > this.exercises.length - 1) {
-                // we are in the last exercise, if out of range "change" active exercise to current in order to trigger a save
-                this.changePage(false, this.exerciseIndex, true);
-            } else {
-                this.changePage(false, newIndex, true);
-            }
+        if (!changeExercise || newIndex > this.exercises.length - 1) {
```

Comment:
I don't understand this logic change here. Comparing it to the previous logic on the left side, I do not understand why it was changed. The PR talks about **hand in early**, which I cannot recognize as part of the code changes.

I suggest to make the code more understandable and not less understandable.

Reply:
Previous logic never calls `changePage` if  `changeExercise` is false. As a result, if `changeExercise` is false, the exercise is not saved because save is triggered by `changePage` method (this part is unintuitive for me as well but as far as I understood the existing implementation works this way). In other words, `saveExercise` did not saved the exercise if `changeExercise` is false, I think this is the actual issue we had here. 

The code called when Hand in Early is clicked is this: https://github.com/ls1intum/Artemis/blob/bugfix/exam-hand-in-early/src/main/webapp/app/exam/participate/exam-navigation-bar/exam-navigation-bar.component.ts#L206-L209. So `handInEarly` does not do more than calling `saveExercise` and emitting.

Reply:
I updated the PR description to point out `saveExercise` is called by `handInEarly` so that it doesn't seem unrelated. I can add a comment to `handInEarly`/`saveExercise` methods if that would help making it more understandable.

Reply:
Maybe writing the method this way can be better, what do you think?

```typescript
saveExercise(changeExercise = true) {
    let newIndex = changeExercise ? this.exerciseIndex + 1 : this.exerciseIndex;
    const submission = ExamParticipationService.getSubmissionForExercise(this.exercises[this.exerciseIndex]);
    // we do not submit programming exercises on a save
    if (submission && this.exercises[this.exerciseIndex].type !== ExerciseType.PROGRAMMING) {
        submission.submitted = true;
    }
    if (newIndex > this.exercises.length - 1) {
        // we are in the last exercise, if out of range "change" active exercise to current in order to trigger a save
        newIndex = this.exerciseIndex;
    }
    this.changePage(false, newIndex, true);
}
```